From 2dff7e827123b3691c20fbe1772cb22ed7c4ea06 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:38:36 -0400
Subject: [virt] kvm/vmx: check use I/O bitmap first before unconditional I/O exit

Message-id: <1406040016-3289-122-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86200
O-Subject: [RHEL7 PATCH v2 121/221] KVM: VMX: check use I/O bitmap first before unconditional I/O exit
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Zhihui Zhang <zzhsuny@gmail.com>

According to Table C-1 of Intel SDM 3C, a VM exit happens on an I/O instruction when
"use I/O bitmaps" VM-execution control was 0 _and_ the "unconditional I/O exiting"
VM-execution control was 1. So we can't just check "unconditional I/O exiting" alone.
This patch was improved by suggestion from Jan Kiszka.

Reviewed-by: Jan Kiszka <jan.kiszka@siemens.com>
Signed-off-by: Zhihui Zhang <zzhsuny@gmail.com>
Signed-off-by: Marcelo Tosatti <mtosatti@redhat.com>
(cherry picked from commit 2f0a6397dd3cac2fb05b46cad08c1d532c04d6b8)
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index e095276..f520142 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -6521,11 +6521,8 @@ static bool nested_vmx_exit_handled_io(struct kvm_vcpu *vcpu,
  int size;
  u8 b;
 
- if (nested_cpu_has(vmcs12, CPU_BASED_UNCOND_IO_EXITING))
-  return 1;
-
  if (!nested_cpu_has(vmcs12, CPU_BASED_USE_IO_BITMAPS))
-  return 0;
+  return nested_cpu_has(vmcs12, CPU_BASED_UNCOND_IO_EXITING);
 
  exit_qualification = vmcs_readl(EXIT_QUALIFICATION);
 
-- 
1.7.1